client/pov-recovery/src/lib.rs

```diff
@@ -423,8 +353,75 @@ where
 			.import_blocks(BlockOrigin::ConsensusBroadcast, incoming_blocks);
 	}
 
+	/// Attempts an explicit recovery of one or more blocks.
+	pub fn recover(&mut self, req: RecoveryRequest<Block>) {
+		let RecoveryRequest { mut hash, delay, kind } = req;
+		let mut to_recover = Vec::new();
+
+		let do_recover = loop {
+			let candidate = match self.candidates.get_mut(&hash) {
+				Some(candidate) if !candidate.waiting_recovery => candidate,
+				None => {
+					tracing::error!(
+						target: LOG_TARGET,
+						block_hash = ?hash,
+						"Cound not recover. Block was never announced as candidate"
+					);
+					break false
+				},
+				// Recovery already in progress
+				_ => break false,
```

Comment:
This branch makes the assumption that for `RecoveryKind::Full` none of the ancestors are in recovery right? Are we sure this assumption is always correct? Lets say we want to recover `C` and its ancestors `A->B->C`. If `A` has `waiting_recovery = true`, we will also not recover `B` and `C`.

Reply:
Good point. Fix in progress